### DISCUSSION TITLE
"Update all projects" Breaks CF Client

### DISCUSSION THREAD

**Discussion by LucasTheIntern** [bitcraze/discussions#932] [Q&A] [answered]
Hello, 

I'm trying to use the "Update all projects" located on the VM's desktop. When I do so, I get the following output:

`Updating /home/bitcraze/projects/crazyflie2-exp-template-electronics
Already up to date.
Updating /home/bitcraze/projects/crazyflie2-nrf-bootloader
Already up to date.
Updating /home/bitcraze/projects/crazyflie2-nrf-firmware
Already up to date.
Updating /home/bitcraze/projects/crazyflie2-stm-bootloader
Already up to date.
Updating /home/bitcraze/projects/crazyflie-android-client
Already up to date.
Updating /home/bitcraze/projects/crazyflie-bootloader
Already up to date.
Updating /home/bitcraze/projects/crazyflie-clients-python
Already up to date.
Updating /home/bitcraze/projects/crazyflie-firmware
Already up to date.
Updating /home/bitcraze/projects/crazyflie-lib-python
Already up to date.
Updating /home/bitcraze/projects/crazyradio-firmware
Already up to date.
Updating /home/bitcraze/projects/lps-node-firmware
Already up to date.
Obtaining file:///home/bitcraze/projects/crazyflie-lib-python
Collecting libusb-package~=1.0
  Using cached libusb_package-1.0.26.2-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (70 kB)
Collecting numpy<1.25,>=1.20
  Using cached numpy-1.24.4-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.3 MB)
Collecting pyusb>=1.0.0b2
  Using cached pyusb-1.2.1-py3-none-any.whl (58 kB)
Collecting scipy~=1.7
  Using cached scipy-1.10.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (34.5 MB)
Collecting importlib-resources
  Using cached importlib_resources-6.1.0-py3-none-any.whl (33 kB)
Collecting zipp>=3.1.0; python_version < "3.10"
  Using cached zipp-3.17.0-py3-none-any.whl (7.4 kB)
ERROR: cfclient 2023.6.post13+16b1b72 requires pyqt6~=6.5, which is not installed.
ERROR: cfclient 2023.6.post13+16b1b72 requires PyQt6-sip~=13.5, which is not installed.
ERROR: cfclient 2023.6.post13+16b1b72 has requirement pyqtgraph~=0.13, but you'll have pyqtgraph 0.12.4 which is incompatible.
ERROR: cfclient 2023.6.post13+16b1b72 has requirement PyYAML~=6.0.1, but you'll have pyyaml 5.3.1 which is incompatible.
ERROR: cfclient 2023.6.post13+16b1b72 has requirement pyzmq~=25.0, but you'll have pyzmq 22.3.0 which is incompatible.
ERROR: cfclient 2023.6.post13+16b1b72 has requirement vispy~=0.13.0, but you'll have vispy 0.9.6 which is incompatible.
Installing collected packages: zipp, importlib-resources, libusb-package, numpy, pyusb, scipy, cflib
  Attempting uninstall: zipp
    Found existing installation: zipp 3.17.0
    Uninstalling zipp-3.17.0:
      Successfully uninstalled zipp-3.17.0
  Attempting uninstall: importlib-resources
    Found existing installation: importlib-resources 6.1.0
    Uninstalling importlib-resources-6.1.0:
      Successfully uninstalled importlib-resources-6.1.0
  Attempting uninstall: libusb-package
    Found existing installation: libusb-package 1.0.26.2
    Uninstalling libusb-package-1.0.26.2:
      Successfully uninstalled libusb-package-1.0.26.2
  Attempting uninstall: numpy
    Found existing installation: numpy 1.24.4
    Uninstalling numpy-1.24.4:
      Successfully uninstalled numpy-1.24.4
  Attempting uninstall: pyusb
    Found existing installation: pyusb 1.2.1
    Uninstalling pyusb-1.2.1:
      Successfully uninstalled pyusb-1.2.1
  Attempting uninstall: scipy
    Found existing installation: scipy 1.10.1
    Uninstalling scipy-1.10.1:
      Successfully uninstalled scipy-1.10.1
  Attempting uninstall: cflib
    Found existing installation: cflib 0.1.23
    Uninstalling cflib-0.1.23:
      Successfully uninstalled cflib-0.1.23
  Running setup.py develop for cflib
Successfully installed cflib importlib-resources-6.1.0 libusb-package-1.0.26.2 numpy-1.24.4 pyusb-1.2.1 scipy-1.10.1 zipp-3.17.0
Obtaining file:///home/bitcraze/projects/crazyflie-clients-python
Collecting PyQt6-sip~=13.5
  Using cached PyQt6_sip-13.5.2-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.whl (298 kB)
Collecting PyYAML~=6.0.1
  Using cached PyYAML-6.0.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (736 kB)
Collecting appdirs~=1.4.0
  Using cached appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)
Collecting cflib>=0.1.23
  Using cached cflib-0.1.23-py2.py3-none-any.whl (188 kB)
Collecting numpy<1.25,>=1.20
  Using cached numpy-1.24.4-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.3 MB)
Collecting pyqt6~=6.5
  Using cached PyQt6-6.5.2.tar.gz (1.0 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... error
    ERROR: Command errored out with exit status 1:
     command: /usr/bin/python3 /tmp/tmp2hlvevd0 prepare_metadata_for_build_wheel /tmp/tmpu9k51y3_
         cwd: /tmp/pip-install-7tz7hii5/pyqt6
    Complete output (36 lines):
    Querying qmake about your Qt installation...
    Traceback (most recent call last):
      File "/tmp/tmp2hlvevd0", line 126, in prepare_metadata_for_build_wheel
        hook = backend.prepare_metadata_for_build_wheel
    AttributeError: module 'sipbuild.api' has no attribute 'prepare_metadata_for_build_wheel'
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "/tmp/tmp2hlvevd0", line 280, in <module>
        main()
      File "/tmp/tmp2hlvevd0", line 263, in main
        json_out['return_val'] = hook(**hook_input['kwargs'])
      File "/tmp/tmp2hlvevd0", line 130, in prepare_metadata_for_build_wheel
        return _get_wheel_metadata_from_wheel(backend, metadata_directory,
      File "/tmp/tmp2hlvevd0", line 159, in _get_wheel_metadata_from_wheel
        whl_basename = backend.build_wheel(metadata_directory, config_settings)
      File "/tmp/pip-build-env-5sic8alh/overlay/lib/python3.8/site-packages/sipbuild/api.py", line 46, in build_wheel
        project = AbstractProject.bootstrap('wheel',
      File "/tmp/pip-build-env-5sic8alh/overlay/lib/python3.8/site-packages/sipbuild/abstract_project.py", line 87, in bootstrap
        project.setup(pyproject, tool, tool_description)
      File "/tmp/pip-build-env-5sic8alh/overlay/lib/python3.8/site-packages/sipbuild/project.py", line 586, in setup
        self.apply_user_defaults(tool)
      File "/tmp/pip-install-7tz7hii5/pyqt6/project.py", line 66, in apply_user_defaults
        super().apply_user_defaults(tool)
      File "/tmp/pip-build-env-5sic8alh/overlay/lib/python3.8/site-packages/pyqtbuild/project.py", line 70, in apply_user_defaults
        super().apply_user_defaults(tool)
      File "/tmp/pip-build-env-5sic8alh/overlay/lib/python3.8/site-packages/sipbuild/project.py", line 237, in apply_user_defaults
        self.builder.apply_user_defaults(tool)
      File "/tmp/pip-build-env-5sic8alh/overlay/lib/python3.8/site-packages/pyqtbuild/builder.py", line 78, in apply_user_defaults
        self._get_qt_configuration()
      File "/tmp/pip-build-env-5sic8alh/overlay/lib/python3.8/site-packages/pyqtbuild/builder.py", line 497, in _get_qt_configuration
        for line in project.read_command_pipe([self.qmake, '-query']):
      File "/tmp/pip-build-env-5sic8alh/overlay/lib/python3.8/site-packages/sipbuild/project.py", line 554, in read_command_pipe
        raise UserException(
    sipbuild.exceptions.UserException
    ----------------------------------------
ERROR: Command errored out with exit status 1: /usr/bin/python3 /tmp/tmp2hlvevd0 prepare_metadata_for_build_wheel /tmp/tmpu9k51y3_ Check the logs for full command output.
Press any key to exit
`

The CF Client shortcut does not work afterwards. When the desktop shortcut icon for the client is double-clicked, the terminal window appears briefly, and then disappears. The client never opens. 

I was able to replicate this issue by creating a new instance of the Bitcraze VM in Oracle/VirtualBox using the "BitcrazeVM_2022.05.ova" version. In the new instance, I first checked to make sure the client worked (it did), and then tried updating all projects, which had the same output and caused the client to fail. 

Any idea why this is happening?

**Comment by krichardsson** [comment, marked as answer]
Yes, this is a problem. What seems to happen is that the latest version of the client uses PyQt6 but something goes wrong when pip tries to install it. I'm not sure exactly why, but I think it is related to the fairly old python version (3.8) and the general complexity of qt.

I crated an issue for this https://github.com/bitcraze/bitcraze-vm/issues/66
In the issue there is a workaround (to downgrade the client code), otherwise I hope to be able to release a new VM in a couple of days that should solve the problem.

Thanks for letting us know!

**Reply by LucasTheIntern** [reply]
It took a few tries to get the commands in the suggested workaround to do their thing, but it's working now. Thank you!

**Reply by knmcguire** [reply]
Hi! I opened up the discussion again as that will remove it from the list, and this discussion will be interesting for other users. Since you were able to resolve your problem on @krichardsson answer, I've selected that as the solution :) It was probably due to the discussion being listed as general and not as Q&A.

Just a note for next time :) but happy you were able to resolve your problem! We hope to release the VM this month so it will work better out of the box.
